FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+  optimizing the process through compilation and caching.
+- Setting up the object stack, including the Call Manager, Invocation
+  Containers, Kernels, and more. This includes setting up the WASM runtime, its
+  configuration, binding syscalls as host-provided functions, passing
+  parameters, handling result values, managing aborts.
+- Serving as a call orchestration layer across actors, through the Call Manager
+  abstraction.
+- Handling syscalls from actors.
+- Interacting with the outer environment (Filecoin client) when necessary.
+- Managing IPLD state I/O. Potentially tracking object lifetime to cooperate
+  with state tree garbage collection.
+
+### Core technical design
+
+This diagram serves as a reference for the technical design of the FVM.
+
+![Technical design](./fip-draft_foundational_fvm/technical-design.png)
+
+#### Machine
+
+The Machine is an instantiation of the FVM on a concrete epoch, over a specific
+state root, ready to take in messages for execution as explained above.
+
+#### Call Manager
+
+The Call Manager coordinates the call stack involved in processing a given message.
+
+#### Invocation Container
+
+The Invocation Container (IC) is the environment/sandbox that runs actor code
+within the context of a single invocation. That is, reentrancy or recursion on
+the same actor will spin off another IC.
+
+The IC is an instance of a WASM engine/runtime fulfilling the FVM-side of the
+contract. The contract consists of:
+
+1. Syscalls available as importable host-provided functions.
+2. Shared types to represent data across the FVM boundary.
+3. Actor-owned memory to exchange data in syscalls.
+4. Transparent gas accounting and automatic execution halting.
+5. (Future) Dynamic linkage of popular WASM modules "blessed" by the community
+   through some governance process, in order to reduce actor bytecode size (e.g.
+   FVM SDKs, IPLD codecs, etc).
+
+Because this contract may change over time, user-deployed actors may specify an
+IC version at [`InitActor#InstallActor` time](#actor-deployment). The strategy
+for deprecating IC versions over time has not been yet determined.
+
+#### Syscalls
+
+Syscalls are functions made available to the actor as WASM imports, allowing it
+to interact with the outer environment, and to call complex operations that are
+more performant to implement in native land (e.g. cryptographic primitives).
+
+#### Kernel
+
+The Kernel is an object created by the Call Manager and injected into the
+Invocation Container. It processes syscalls, and stores state concerning the
+invocation.
+
+#### Externs
+
+Similar to syscalls being functions provided to the actor by the FVM, Externs
+are functions provided by the node to the FVM. Depending on the programming
+languages at play, Externs may need to traverse an FFI boundary.
+
+### Actors
+
+The term _Actor_ is a reference to the [actor model], a concurrent computation
+paradigm that inspires Filecoin's programming and scalability philosophies.
```

Comment:
I... wouldn't mention the actor model here given that we've basically ignored it entirely. I'd just say that an actor is equivalent to a smart contract.